packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -815,14 +920,24 @@ class Navigator extends StatefulWidget {
   /// A list of observers for this navigator.
   final List<NavigatorObserver> observers;
 
-  /// The default name for the [initialRoute].
+  /// The name for the default route of the application.
   ///
   /// See also:
   ///
   ///  * [dart:ui.Window.defaultRouteName], which reflects the route that the
   ///    application was started with.
   static const String defaultRouteName = '/';
 
+  /// Called when the widget is created to generate the initial list of [Route]
+  /// objects, if [initialRoute] is not null.
+  ///
+  /// Defaults to [defaultGenerateInitialRoutes].
+  ///
+  /// The [NavigatorState] and [initialRoute] will be passed to the callback.
+  /// The callback must return a list of [Route] objects with which the history
+  /// will be primed.
+  final RouteListFactory onGenerateInitialRoutes;
```

Comment:
@goderbauer I feel tedious when trying to write the migrating doc for break change email. I would like to discuss some change to the current api. From the majority internal test failures, people are trying to customize route generating based on the isInitialRoute. For example
```dart
MaterialApp(
  ...
  onGenerateRoute: (RouteSettings settings) {
     if(settings.isInitialRoutes)
      return Route1()
    else
      return Route2()
  }
)
```
The thing they try to achieve is, given the same route name, they want to render different routes when this route is generating on initial route or generating by a push.

I found this is difficult to migrate becuse two reason
1, defaultGenerateInitialRoutes is a bit complicate, it require both parsing the string and feed them into onGenerateRoute. User will likely need to duplicate a lot of code in defaultGenerateInitialRoutes to get what they want.
2, This api is not surface to WidgetApp, where majority of the users do not directly initiate navigator. It make them hard to provide onGenerateInitialRoutes.

What I am thinking is we should separate defaultGenerateInitialRoutes into a static function called defaultInitialRouteParser. and defaultGenerateInitialRoutes will use the parser to parse the initialroutename and feed that into ongenerateroute. User who override onGenerateInitialRoutes can then use the defaultInitialRouteParser instead of duplicating code.

We will probably also need to surface onGenerateInitialRoutes to material app or cupertino app so that user can really use it for migration.

Reply:
Talked in person to resolve this.